File: worker/src/store/lists.ts
import { newList } from '../lib/ids';
import {
  AlreadyExistsError,
  List,
  ListDetails,
  ListSummary,
  NotFoundError,
  isUniqueViolation,
  nowISO,
} from './types';

export type NewListParams = {
  slug: string;
  name: string;
  company_id: string;
  sending_domain: string;
  description?: string;
  weight?: number;
};

const LIST_SELECT = `SELECT id, slug, name, COALESCE(description, '') AS description,
       COALESCE(weight, 10) AS weight, COALESCE(company_id, '') AS company_id,
       sending_domain, created_at, updated_at FROM lists`;

export async function createList(db: D1Database, p: NewListParams): Promise<List> {
  const id = newList();
  const now = nowISO();
  const weight = p.weight ?? 10;
  const description = p.description ?? '';
  try {
    await db
      .prepare(
        `INSERT INTO lists (id, slug, name, description, weight, company_id, sending_domain, created_at, updated_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      )
      .bind(id, p.slug, p.name, description, weight, p.company_id, p.sending_domain, now, now)
      .run();
  } catch (err) {
    if (isUniqueViolation(err)) throw new AlreadyExistsError();
    throw err;
  }
  return getListByID(db, id);
}

export async function getListByID(db: D1Database, id: string): Promise<List> {
  const row = await db.prepare(`${LIST_SELECT} WHERE id = ?`).bind(id).first<List>();
  if (!row) throw new NotFoundError();
  return row;
}

export async function getListBySlug(db: D1Database, slug: string): Promise<List> {
  const row = await db.prepare(`${LIST_SELECT} WHERE slug = ?`).bind(slug).first<List>();
  if (!row) throw new NotFoundError();
  return row;
}

export async function resolveList(db: D1Database, idOrSlug: string): Promise<List> {
  try {
    return await getListByID(db, idOrSlug);
  } catch {
    return await getListBySlug(db, idOrSlug);
  }
}

export async function listLists(db: D1Database): Promise<ListSummary[]> {
  const { results } = await db
    .prepare(
      `SELECT l.id, l.slug, l.name, COALESCE(l.description, '') AS description,
              COALESCE(l.weight, 10) AS weight, COALESCE(l.company_id, '') AS company_id,
              l.sending_domain, l.created_at, l.updated_at,
              COALESCE(SUM(CASE WHEN subs.subscribed = 1 THEN 1 ELSE 0 END), 0) AS subscribed_count
         FROM lists l
         LEFT JOIN subscriptions subs ON subs.list_id = l.id
         GROUP BY l.id
         ORDER BY l.weight ASC, l.name ASC`,
    )
    .all<ListSummary>();
  return results;
}

export async function getListDetails(db: D1Database, listID: string): Promise<ListDetails> {
  const row = await db
    .prepare(
      `SELECT l.id, l.slug, l.name, COALESCE(l.description, '') AS description,
              COALESCE(l.weight, 10) AS weight, COALESCE(l.company_id, '') AS company_id,
              l.sending_domain, l.created_at, l.updated_at,
              COALESCE(SUM(CASE WHEN subs.subscribed = 1 THEN 1 ELSE 0 END), 0) AS subscribed_count,
              COUNT(subs.id) AS total_count
         FROM lists l
         LEFT JOIN subscriptions subs ON subs.list_id = l.id
        WHERE l.id = ?
        GROUP BY l.id`,
    )
    .bind(listID)
    .first<ListSummary & { total_count: number }>();
  if (!row || !row.id) throw new NotFoundError();
  const lastSend = await db
    .prepare('SELECT MAX(created_at) AS last_send_at FROM sends WHERE list_id = ?')
    .bind(listID)
    .first<{ last_send_at: string | null }>();
  return { ...row, last_send_at: lastSend?.last_send_at ?? null };
}

File: src/internal/store/lists.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"fmt"

	"github.com/ranaroussi/minigun/internal/ids"
	"github.com/ranaroussi/minigun/internal/models"
)

var ErrNotFound = errors.New("not found")
var ErrAlreadyExists = errors.New("already exists")

type NewListParams struct {
	Slug          string
	Name          string
	CompanyID     string
	SendingDomain string
	Description   string
	Weight        int
}

func (s *Store) CreateList(ctx context.Context, p NewListParams) (*models.List, error) {
	id := ids.NewList()
	now := nowISO()
	if p.Weight == 0 {
		p.Weight = 10
	}
	_, err := s.DB.ExecContext(ctx,
		`INSERT INTO lists (id, slug, name, description, weight, company_id, sending_domain, created_at, updated_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		id, p.Slug, p.Name, p.Description, p.Weight, p.CompanyID, p.SendingDomain, now, now,
	)
	if err != nil {
		if isUniqueViolation(err) {
			return nil, ErrAlreadyExists
		}
		return nil, fmt.Errorf("insert list: %w", err)
	}
	return s.GetListByID(ctx, id)
}

const listSelect = `SELECT id, slug, name, COALESCE(description, '') AS description,
		COALESCE(weight, 10) AS weight, COALESCE(company_id, '') AS company_id,
		sending_domain, created_at, updated_at FROM lists`

func (s *Store) GetListByID(ctx context.Context, id string) (*models.List, error) {
	return s.queryList(ctx, listSelect+` WHERE id = ?`, id)
}

func (s *Store) GetListBySlug(ctx context.Context, slug string) (*models.List, error) {
	return s.queryList(ctx, listSelect+` WHERE slug = ?`, slug)
}

func (s *Store) ResolveList(ctx context.Context, idOrSlug string) (*models.List, error) {
	if l, err := s.GetListByID(ctx, idOrSlug); err == nil {
		return l, nil
	}
	return s.GetListBySlug(ctx, idOrSlug)
}

func (s *Store) queryList(ctx context.Context, q string, args ...any) (*models.List, error) {
	row := s.DB.QueryRowContext(ctx, q, args...)
	var l models.List
	var created, updated string
	if err := row.Scan(&l.ID, &l.Slug, &l.Name, &l.Description, &l.Weight, &l.CompanyID, &l.SendingDomain, &created, &updated); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, err
	}
	var err error
	if l.CreatedAt, err = parseTime(created); err != nil {
		return nil, err
	}
	if l.UpdatedAt, err = parseTime(updated); err != nil {
		return nil, err
	}
	return &l, nil
}

func isUniqueViolation(err error) bool {
	if err == nil {
		return false
	}
	msg := err.Error()
	return contains(msg, "UNIQUE constraint failed") || contains(msg, "constraint failed: UNIQUE")
}

func contains(s, sub string) bool {
	return len(sub) > 0 && len(s) >= len(sub) && indexOf(s, sub) >= 0
}

func indexOf(s, sub string) int {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return i
		}
	}
	return -1
}
